pytorch_lightning/cluster_environments/lsf_environment.py

```diff
@@ -0,0 +1,176 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import re
+import socket
+import warnings
+from pytorch_lightning import _logger as log
+from pytorch_lightning.cluster_environments.cluster_environment import ClusterEnvironment
+
+
+class LSFEnvironment(ClusterEnvironment):
+    """An environment for running on clusters managed by the LSF resource manager.
+
+    It is expected that any execution using this ClusterEnvironment was executed
+    using the Job Step Manager i.e. jsrun.
+
+    This plugin expects the following environment variables:
+
+    LSB_JOBID
+      The LSF assigned job ID
+
+    LSB_HOSTS
+      The hosts used in the job. This string is expected to have the format "batch <rank_0_host> ...."
+
+    JSM_NAMESPACE_LOCAL_RANK
```

Comment:
This is a Coral based LSF integration and not a native LSF integration.  So, we need to be clear that LSF does not include JSM today.  To create a native LSF integration, this adaptation will have to change.  It may not be a major change, but someone will have to issue a follow-on integration to make that happen.

Reply:
What is Coral?  I determined these variables by printing the environment from a job i.e. they were not documented anywhere. Are you aware of any environment variables set by LSF that gives rank and size? I've only been able to find info [here](https://www.ibm.com/support/knowledgecenter/SSWRJV_10.1.0/lsf_config_ref/lsf_envars_job_exec.html)

Reply:
The Coral deployment is a deployment on IBM Power systems that include JSM "Job Step Manager".  To get rank and size, you can use:

`LSB_MCPU_HOSTS` which follow as such "hostnameA num_ranks hostnameB num_ranks ..."
`LSB_DJOB_RANKFIKLE` which includes and OpenMPI compatible rankfile format stating specifically which cores/threads to bind to on reach host.

Of course the GPU's to use are in both the CUDA and NVIDIA environment variables.  Inside a standard job, one contained in a control group, you can only see the GPU's that you requested, and they all start with GPU0, so regardless of the actual GPU's you have allocated, they start 0, 1, 2, ... inside the control group.